corgidrp/l4_to_tda.py

```diff
@@ -0,0 +1,73 @@
+# A file that holds the functions that transmogrify l4 data to TDA (Technical Demo Analysis) data 
+import corgidrp.fluxcal as fluxcal
+
+def determine_color_cor(input_dataset, calspec_filepath, source_filepath):
+    """
+    determine the color correction factor of the observed source
+    at the reference wavelength of the used filter band and put it into the header.
+    We assume that each frame in the dataset was observed with the same color filter.
+    
+    Args:
+        input_dataset (corgidrp.data.Dataset): a dataset of Images (L2b-level)
+        calspec_filepath (str): file name of the known reference flux (usually CALSPEC)
+        source_filepath (str): file name of the flux model of the observed source in 
+                               CALSPEC units (erg/(s * cm^2 * AA) and format
+    
+    Returns:
+        corgidrp.data.Dataset: a version of the input dataset with updated header including 
+                              the reference wavelength and the color correction factor
+    """
+    color_dataset = input_dataset.copy()
+    # get the filter name from the header keyword 'CFAMNAME'
+    filter_name = fluxcal.get_filter_name(color_dataset)
+    # read the transmission curve from the color filter file
+    wave, filter_trans = fluxcal.read_filter_curve(filter_name)
+    # calculate the reference wavelength of the color filter
+    lambda_ref = fluxcal.calculate_pivot_lambda(filter_trans, wave)
+    # calculate the flux from the user given CALSPEC file binned on the wavelength grid of the filter
+    flux_ref = fluxcal.read_cal_spec(calspec_filepath, wave)
+    # we assume that the source spectrum is a calspec standard or its 
+    # model data is in a file with the same format and unit as the calspec data
+    source_sed = fluxcal.read_cal_spec(source_filepath, wave)
+    #Calculate the color correction factor
+    k = fluxcal.compute_color_cor(filter_trans, wave, flux_ref, lambda_ref, source_sed)
+    
+    # write the reference wavelength and the color correction factor to the header (keyword names tbd)
+    history_msg = "the color correction is calculated and added to the header {0}".format(str(k))
```

Comment:
@JuergenSchreiber and @semaphoreP - so the color correction might change slightly with different star spectral types - is it worth putting in the cal star's spectral type into the header as well or leave that as an exercise for the user?

Also: just want to be abundantly clear on this - the true purpose of these absolute flux calibrations is to calibrate out the ND filter. CGI will be delivering the flux of the planet relative to the flux of the star - this greatly simplifies our science calibration error budget as it puts everything into relative terms, instead of absolute.

We need the ND filter to observe the unocculted flux of the exoplanet host star. We observe our planet/disk without the ND filter. So, the difference between observations of the star vs. those of the planet/disk is that ND filter. We need to effectively measure the ND filter's response in order to "convert" our unocculted star measurements with the ND filter to remove the ND filter's response. Also, the ND filter is in a focal plane, so depositions can occur on the ND filter and change the spectral response- and this can vary by the position on the ND filter.

The way we came up with removing/correcting for the ND filter is to perform absolute flux calibrations with and without the ND filter in place. At the end of the day, we just need to remove that silly ND filter from our measurements of the unocculted host star.

Reply:
@rzellem - thanks for further clarifying. I am aware of that and that simplifies things, e.g. we do not have to care about units, unit conversions etc. My way forward would be to determine the calibration factor from a flux standard observed without ND, then observe a suitable other (ND) star with known flux in the same band with the ND. Applying the color corrected calibration factor to the DNs of that ND star would yield the measured ND flux. Dividing the measured ND flux by the known ND star flux would yield the ND transmission (assuming a sufficient constant ND transmission within the color filter band).
The spectral type of the e.g. calspec stars is listed in the calspec table on stsci.edu, so I think the target name of the observed stars in the header would be sufficient enough.

Reply:
@JuergenSchreiber - apologies for not making this clearer sooner! There's certainly nothing wrong with what you currently have - and honestly, it will probably be the best thing in the long term in case someone who does an observation of a system (maybe a disk?) does want absolute flux units - so I think your code is ultimately for the better, thank you!

We found that in order to achieve our allocation in the Science Calibration error budget, that we had to observe multiple stars to increase our SNR - a single CALSPEC standard did not offer enough precision. If you take a look at Linsey's poster (and we're working on getting you her code - but it needs to be blessed by JPL first), she had a few dim stars (~4) to observe without the ND filter and then another ~6-10 bright ones to observe with the ND filter. By combining all of these measurements together, she was able to determine the correction factor with the required precision.

Reply:
@rzellem, I think we can just re-run the abs phot calib and/or ND filter calib on multiple reduced datasets after we process all the individual datasets first (e.g., the color corrections are already in the headers). So I don't think this step precludes doing multiple stars like you mentioned.

Reply:
However, I think if the code in this function is quick to run, we might as well run it on a per-file basis (see my above comment).

Reply:
@semaphoreP, I agree. I assume that we run the complete pipeline to abs flux cal for each observation of one star in one filter band. And for the color correction I assume, that we know already what the reference star is (should be in the abs flux cal file). For the reference star observation itself and determination of the calibration factor which is a cal file e2e processing no color correction is needed.  In the cal file there should be in the end  calibration factors of reference stars with and without ND filter. Well, I think we will see how to organize this in the course of the development. 
For the ND filter calibration, it is getting more complex, in the end we will have some combinations of bright with dim cal standard stars reduced, on which we determine the filter transmission. We have to see how we organize this within the pipeline. Inside a dataset, or set up a dictionary, maybe something like the associations in JWST pipeline at this higher level stage of processing, just brainstorming.

Reply:
This makes sense to me. I agree that it'll be best to write the abs flux cal (and then the ND calib), and see if there's anything missing in this function as we do that. Shall we wrap up this PR and move on to that?